File: apiClient.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;
using System.Text.Json;

namespace FaPiaoApiClient
{
    class Program
    {
        // 配置信息
        private static readonly string BaseUrl = "https://api.fa-piao.com";
        private static readonly string AppKey = "YOUR_APP_KEY"; // 替换为您的AppKey
        private static readonly string AppSecret = "YOUR_SECRET_KEY"; // 替换为您的密钥

        static async Task Main(string[] args)
        {
            try
            {
                // 准备请求头参数
                string method = "POST";
                string path = "/v5/enterprise/authorization";
                
                // 请求参数
                var formData = new Dictionary<string, string>
                {
                    { "nsrsbh", "915101820724315989" }
                };

                string randomString = GenerateRandomString(20);
                string timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds().ToString();
                
                // 计算签名
                string signature = CalculateSignature(method, path, randomString, timestamp);
                
                // 发送请求
                var result = await SendRequest(BaseUrl + path, formData, method, randomString, timestamp, signature);
                
                // 输出结果
                Console.WriteLine($"{method}: {BaseUrl}{path}");
                Console.WriteLine($"headers: {GetHeadersJson(randomString, timestamp, signature)}");
                Console.WriteLine($"form-data: {JsonSerializer.Serialize(formData)}");
                Console.WriteLine($"HTTP状态码: {result.StatusCode}");
                Console.WriteLine($"响应内容: {result.Content}");
                
                // 尝试解析JSON响应
                try
                {
                    var responseObj = JsonSerializer.Deserialize<object>(result.Content);
                    Console.WriteLine("解析后的响应:");
                    Console.WriteLine(JsonSerializer.Serialize(responseObj, new JsonSerializerOptions { WriteIndented = true }));
                }
                catch (JsonException)
                {
                    // 响应不是有效的JSON格式
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"发生错误: {ex.Message}");
            }
        }

        /// <summary>
        /// 生成随机字符串
        /// </summary>
        private static string GenerateRandomString(int length)
        {
            const string chars = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
            var random = new Random();
            var result = new StringBuilder(length);
            
            for (int i = 0; i < length; i++)
            {
                result.Append(chars[random.Next(chars.Length)]);
            }
            
            return result.ToString();
        }

        /// <summary>
        /// 计算HMAC-SHA256签名
        /// </summary>
        private static string CalculateSignature(string method, string path, string randomString, string timestamp)
        {
            // 构建签名字符串，与服务器端保持一致
            string signContent = $"Method={method}&Path={path}&RandomString={randomString}&TimeStamp={timestamp}&AppKey={AppKey}";
            
            // 使用HMAC-SHA256计算签名，以secret作为密钥
            using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(AppSecret)))
            {
                byte[] hashBytes = hmac.ComputeHash(Encoding.UTF8.GetBytes(signContent));
                return BitConverter.ToString(hashBytes).Replace("-", "").ToUpper();
            }
        }

        /// <summary>
        /// 发送HTTP请求
        /// </summary>
        private static async Task<(int StatusCode, string Content)> SendRequest(
            string url, Dictionary<string, string> formData, string method, 
            string randomString, string timestamp, string signature)
        {
            using (var httpClient = new HttpClient())
            {
                // 设置请求头
                httpClient.DefaultRequestHeaders.Add("AppKey", AppKey);
                httpClient.DefaultRequestHeaders.Add("Sign", signature);
                httpClient.DefaultRequestHeaders.Add("TimeStamp", timestamp);
                httpClient.DefaultRequestHeaders.Add("RandomString", randomString);
                
                // 创建multipart/form-data内容
                using (var content = new MultipartFormDataContent())
                {
                    foreach (var item in formData)
                    {
                        content.Add(new StringContent(item.Value), item.Key);
                    }
                    
                    // 发送请求
                    var response = await httpClient.PostAsync(url, content);
                    
                    // 读取响应
                    string responseContent = await response.Content.ReadAsStringAsync();
                    
                    return ((int)response.StatusCode, responseContent);
                }
            }
        }

        /// <summary>
        /// 获取请求头JSON字符串
        /// </summary>
        private static string GetHeadersJson(string randomString, string timestamp, string signature)
        {
            var headers = new Dictionary<string, string>
            {
                { "AppKey", AppKey },
                { "Sign", signature },
                { "TimeStamp", timestamp },
                { "RandomString", randomString },
                { "Content-Type", "multipart/form-data" }
            };
            
            return JsonSerializer.Serialize(headers);
        }
    }
}

File: apiClient.js
/**
 * 数电发票API
 * 获取授权 客户端签名验证示例
 * 请求接口: https://api.fa-piao.com/v5/enterprise/authorization
 */

const crypto = require('crypto');
const https = require('https');
const querystring = require('querystring');
const FormData = require('form-data');

// 配置信息
const config = {
    baseUrl: 'https://api.fa-piao.com',
    appKey: 'YOUR_APP_KEY', // 替换为您的AppKey
    appSecret: 'YOUR_SECRET_KEY', // 替换为您的密钥
};

/**
 * 生成随机字符串
 * @param {number} length 字符串长度
 * @returns {string} 随机字符串
 */
function generateRandomString(length = 16) {
    const characters = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
    let randomString = '';
    for (let i = 0; i < length; i++) {
        randomString += characters.charAt(Math.floor(Math.random() * characters.length));
    }
    return randomString;
}

/**
 * 计算签名
 * @param {string} method HTTP方法
 * @param {string} path 请求路径
 * @param {string} randomString 随机字符串
 * @param {string} timestamp 时间戳
 * @param {string} appKey 应用密钥
 * @param {string} appSecret 应用密钥
 * @returns {string} 签名
 */
function calculateSignature(method, path, randomString, timestamp, appKey, appSecret) {
    // 构建签名字符串，与服务器端保持一致
    const signContent = `Method=${method}&Path=${path}&RandomString=${randomString}&TimeStamp=${timestamp}&AppKey=${appKey}`;
    
    // 使用HMAC-SHA256计算签名，以secret作为密钥
    const hmac = crypto.createHmac('sha256', appSecret);
    hmac.update(signContent);
    const signature = hmac.digest('hex');
    
    // 转为大写
    return signature.toUpperCase();
}

/**
 * 发送请求
 * @param {string} url 请求URL
 * @param {object} formData 表单数据
 * @param {object} headers 请求头
 * @returns {Promise<object>} 响应结果
 */
function sendRequest(url, formData, headers) {
    return new Promise((resolve, reject) => {
        // 创建form-data
        const form = new FormData();
        for (const key in formData) {
            form.append(key, formData[key]);
        }
        
        // 获取form-data的headers
        const formHeaders = form.getHeaders();
        
        // 解析URL
        const urlObj = new URL(url);
        
        // 请求选项
        const options = {
            hostname: urlObj.hostname,
            port: urlObj.port || 443,
            path: urlObj.pathname,
            method: 'POST',
            headers: {
                ...formHeaders,
                ...headers
            },
            rejectUnauthorized: false // 开发环境可设置为false，生产环境建议设置为true
        };
        
        // 删除Content-Type，使用form-data自动生成的
        delete options.headers['Content-Type'];
        
        const req = https.request(options, (res) => {
            let data = '';
            
            res.on('data', (chunk) => {
                data += chunk;
            });
            
            res.on('end', () => {
                resolve({
                    http_code: res.statusCode,
                    response: data
                });
            });
        });
        
        req.on('error', (error) => {
            console.error('请求错误:', error);
            reject(error);
        });
        
        // 发送form-data
        form.pipe(req);
    });
}

/**
 * 主函数
 */
async function main() {
    try {
        // 准备请求头参数
        const method = 'POST';
        const path = '/v5/enterprise/authorization';
        
        // 请求参数
        const formData = {
            nsrsbh: '915101820724315989'
        };

        const randomString = generateRandomString(20);
        const timestamp = Math.floor(Date.now() / 1000).toString(); // 当前时间戳
        
        // 计算签名
        const signature = calculateSignature(
            method, 
            path, 
            randomString, 
            timestamp, 
            config.appKey, 
            config.appSecret
        );
        
        // 设置请求头
        const headers = {
            'AppKey': config.appKey,
            'Sign': signature,
            'TimeStamp': timestamp,
            'RandomString': randomString
        };
        
        // 发送请求
        const result = await sendRequest(config.baseUrl + path, formData, headers);
        
        // 输出结果
        console.log(`${method}: ${config.baseUrl}${path}`);
        console.log(`headers: ${JSON.stringify(headers)}`);
        console.log(`form-data: ${JSON.stringify(formData)}`);
        console.log(`HTTP状态码: ${result.http_code}`);
        console.log(`响应内容: ${result.response}`);
        
    } catch (error) {
        console.error(`发生错误: ${error.message}`);
    }
}

// 执行主函数
main();
